packages/ember-metal/lib/array.js

```diff
@@ -1,7 +1,141 @@
-export function objectAt(content, idx) {
-  if (typeof content.objectAt === 'function') {
-    return content.objectAt(idx);
+import { notifyPropertyChange } from "./property_events";
+import { eachProxyArrayDidChange, eachProxyArrayWillChange } from "./each_proxy";
+import { peekMeta } from "./meta";
+import { sendEvent, removeListener, addListener } from "./events";
+import { peekCacheFor } from "./computed";
+import { get } from "./property_get";
+
+export function objectAt(array, index) {
+  if (Array.isArray(array)) {
+    return array[index];
   } else {
-    return content[idx];
+    return array.objectAt(index);
   }
 }
+
+const EMPTY_ARRAY = [];
+
+export function replace(array, start, deleteCount, items = EMPTY_ARRAY) {
+  if (Array.isArray(array)) {
+    replaceInNativeArray(array, start, deleteCount, items);
+  } else {
+    array.replace(start, deleteCount, items);
+  }
+}
+
+const CHUNK_SIZE = 60000;
+
+// To avoid overflowing the stack, we splice up to CHUNK_SIZE items at a time.
+// See https://code.google.com/p/chromium/issues/detail?id=56588 for more details.
+export function replaceInNativeArray(array, start, deleteCount, items) {
+  arrayContentWillChange(array, start, deleteCount, items.length);
+
+  if (items.length <= CHUNK_SIZE) {
+    array.splice(start, deleteCount, ...items);
+  } else {
+    array.splice(start, deleteCount);
+
+    for (let i = 0; i < items.length; i += CHUNK_SIZE) {
+      let chunk = items.slice(i, i + CHUNK_SIZE);
+      array.splice(start + i, 0, ...chunk);
+    }
+  }
```

Comment:
Most of this file is a direct copy and paste from packages/ember-runtime/lib/mixins/array.js, although I did rewrite `replaceInNativeArray` to be a bit simpler.